sys/net/application_layer/rdcli/rdcli.c

```diff
@@ -0,0 +1,271 @@
+
+/* allocate the default UDP endpoint for reaching RD servers */
+static const sock_udp_ep_t default_remote = {
+    .family    = AF_INET6,
+    .netif     = SOCK_ADDR_ANY_NETIF,
+    .addr      = RDCLI_SERVER_ADDR ,
+    .port      = RDCLI_SERVER_PORT
+};
+
+static size_t parse_location_path(coap_pkt_t *pdu)
+{
+    size_t pos = 0;
+    coap_opt_t opt;
+    uint8_t *tmp = pdu->options;
+
+    tmp = coap_find_opt(pdu, tmp, &opt, COAP_OPT_LOCATION_PATH);
+    if (!tmp) {
+        return 0;
+    }
+    do {
+        if ((pos + opt.len) >= RDCLI_RES_LOC_LEN) {
+            return 0;
+        }
+
+        /* save part of location */
+        active_rd.res_loc[pos++] = '/';
+        memcpy(active_rd.res_loc + pos, opt.val, opt.len);
+        pos += opt.len;
+
+        /* try to find more location options */
+        tmp = coap_find_opt(pdu, tmp, &opt, 0);
+    } while (tmp);
+
+    return pos;
+}
+
+static void on_register(unsigned req_state, coap_pkt_t* pdu, sock_udp_ep_t *remote)
+{
+    thread_flags_t flag = FLAG_ERR;
+
+    if ((req_state == GCOAP_MEMO_RESP) && (pdu->hdr->code == COAP_CODE_CREATED)) {
+        /* read the location header and save the RD details on success */
+        if (parse_location_path(pdu) > 0) {
+            memcpy(&active_rd.remote, remote, sizeof(active_rd.remote));
+            flag = FLAG_SUCCESS;
+        }
+        else {
+            DEBUG("on_register: unable to parse location path\n");
+            /* reset RD entry */
+            active_rd.res_loc[0] = 0;
+        }
+    }
+    else if (req_state == GCOAP_MEMO_TIMEOUT) {
+        DEBUG("on_register: timeout while trying to reach the RD\n");
+        flag = FLAG_TIMEOUT;
+    }
+
+    thread_flags_set((thread_t *)t_waiting, flag);
+}
+
+static void on_update_remove(unsigned req_state, coap_pkt_t *pdu, uint8_t code)
+{
+    thread_flags_t flag = FLAG_ERR;
+
+    if ((req_state == GCOAP_MEMO_RESP) && (pdu->hdr->code == code)) {
+        flag = FLAG_SUCCESS;
+    }
+    else if (req_state == GCOAP_MEMO_TIMEOUT) {
+        flag = FLAG_TIMEOUT;
+    }
+
+    thread_flags_set((thread_t *)t_waiting, flag);
+}
+
+static void on_update(unsigned req_state, coap_pkt_t *pdu, sock_udp_ep_t *remote)
+{
+    (void)remote;
+    on_update_remove(req_state, pdu, COAP_CODE_CHANGED);
+}
+
+static void on_remove(unsigned req_state, coap_pkt_t *pdu, sock_udp_ep_t *remote)
+{
+    (void)remote;
+    on_update_remove(req_state, pdu, COAP_CODE_DELETED);
+}
+
+int rdcli_register(sock_udp_ep_t *remote)
+{
+    int retval = RDCLI_OK;
+    coap_pkt_t pkt;
+
+    /* fall back to the default RD server address if none was given */
+    if (!remote) {
+        remote = (sock_udp_ep_t *)&default_remote;
+    }
+
+    mutex_lock(&lock);
+
+    /* reset existing RD entry */
+    active_rd.res_loc[0] = 0;
+
+    /* build CoAP request packet */
+    int res = gcoap_req_init(&pkt, buf, sizeof(buf), COAP_METHOD_POST, "/rd");
+    if (res < 0) {
+        goto rdcli_register;
```

Comment:
AFAIK not in general, and though I am still not a huge fan of this, I think for certain error handling cases it can make the code more readable, if used with caution...